python/ray/data/_internal/stage_impl.py

```diff
@@ -147,50 +150,90 @@ class ZipStage(AllToAllStage):
     """Implementation of `Dataset.zip()`."""
 
     def __init__(self, other: "Dataset"):
-        def do_zip_all(block_list, clear_input_blocks: bool, *_):
-            blocks1 = block_list.get_blocks()
-            blocks2 = other.get_internal_block_refs()
+        def do_zip_all(block_list: BlockList, clear_input_blocks: bool, *_):
+            # Repartition other to align with this dataset, and then zip together the
+            # blocks in parallel.
+            # TODO(Clark): Port this to a streaming zip, e.g. push block pairs through
+            # an actor that buffers and zips.
+            blocks_with_metadata = block_list.get_blocks_with_metadata()
+            # Get the split indices that will align other with self.
+            block_rows = _calculate_blocks_rows(blocks_with_metadata)
+            indices = list(itertools.accumulate(block_rows))
+            indices.pop(-1)
+
+            # Execute other to a block list.
+            other_block_list = other._plan.execute()
+            other_blocks_with_metadata = other_block_list.get_blocks_with_metadata()
+            other_block_rows = _calculate_blocks_rows(other_blocks_with_metadata)
+            # Check that each dataset has the same number of rows.
+            # TODO(Clark): Support different number of rows via user-directed
+            # dropping/padding.
+            if sum(block_rows) != sum(other_block_rows):
+                raise ValueError("Cannot zip datasets of different number of rows.")
+
+            # Split other at the alignment indices, such that for every block in
+            # block_list, we have a list of blocks from other that has the same
+            # cumulative number of rows as that block.
+            # NOTE: _split_at_indices has a no-op fastpath if the blocks are already
+            # aligned.
+            aligned_other_blocks_with_metadata = _split_at_indices(
```

Comment:
Could it be more efficient to split the dataset that has more blocks? Intuitively it seems tend to have less number of blocks to be split.

Reply:
I think we should split the dataset so that it minimizes data copy in `split_at_indices()`. btw I feel a simple indicator is to split the dataset whose size is smaller.

Reply:
Yep I'll change this to split the smaller (in bytes) dataset to start, we can later add an optimization pass that minimizes splits.